[tflite] add Hexagon delegate to label_image

freedomtan opened the pull request:
add Hexagon Delegte support to label_image and remove unnecessary
depependency in //tensorflow/lite/examples/label_image/BUILD

freedomtan committed:
add Hexagon delegate to label_image

add Hexagon Delegte support to label_image and remove unnecessary
depependency in //tensorflow/lite/examples/label_image/BUILD

karimnosseir commented on tensorflow/lite/examples/label_image/BUILD:
```diff
@@ -35,11 +35,7 @@ cc_binary(
         "//tensorflow/lite/profiling:profiler",
         "//tensorflow/lite/tools/evaluation:utils",
         "@com_google_absl//absl/memory",
-    ] + select({
```
Why are you removing the gpu delegate dependency ?

freedomtan replied:
Because `//tensorflow/lite/tools/evaluation:utils` includes gpu delegate dependency already. Those gpu delegate dependency lines were added before using  `//tensorflow/lite/tools/evaluation:utils`.

karimnosseir replied:
Each build rule should keep the dependency on what it is directly uses. Relying on indirect dependency can make things break easily.
Please put the direct dependency
thanks

freedomtan replied:
OK, I'll do it.

karimnosseir commented on tensorflow/lite/examples/label_image/label_image.cc:
```diff
@@ -361,13 +374,14 @@ int Main(int argc, char** argv) {
         {"max_profiling_buffer_entries", required_argument, nullptr, 'e'},
         {"warmup_runs", required_argument, nullptr, 'w'},
         {"gl_backend", required_argument, nullptr, 'g'},
+        {"hexagon_delegate", required_argument, nullptr, 'j'},
```
This is should be optional ? Why required argument ?

freedomtan replied:
required_argument means the optional requires an argument  (`-j 0|1` instead of `-j`), just like what nnapi (`-a 0|1`) and gpu (`-g 0|1`) delegates do.

karimnosseir reviewed (commented):
Update documentation in 
https://github.com/tensorflow/tensorflow/tree/master/tensorflow/lite/examples/label_image

to explain what is needed to run using hexagon.

Thanks

freedomtan commented:
> Update documentation in
> https://github.com/tensorflow/tensorflow/tree/master/tensorflow/lite/examples/label_image
> 
> to explain what is needed to run using hexagon.
> 
> Thanks

freedomtan committed:
update tflite label_image doc

karimnosseir commented on tensorflow/lite/examples/label_image/README.md:
```diff
@@ -78,4 +78,86 @@ average time: 68.12 ms
+
+To run a model with the Hexagon Delegate, assuming we have
+followed the
+[Hexagon Delegate doc](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/g3doc/performance/hexagon_delegate.md)
```
nit: Hexagon Delegate Guide

karimnosseir commented on tensorflow/lite/examples/label_image/README.md:
````diff
@@ -78,4 +78,86 @@ average time: 68.12 ms
+[Hexagon Delegate doc](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/g3doc/performance/hexagon_delegate.md)
+and installed Hexagon libraries in `/data/local/tmp`. Run it
+```
+adb shell "export LD_LIBRARY_PATH=/data/local/tmp/; \
````
We can get rid of the "export LD_LIBRARY_PATH=/data/local/tmp/" by adding rpath=/data/local/tmp/ to the linkopts in the build rule.
Like here
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/tools/benchmark/BUILD#L32

freedomtan committed:
address review concerns

1. add gpu_delegate depenedency back
2. add hexagon_delegate dependency
3. add -rpath for hexagon delegate
3. modify readme

freedomtan committed:
fix a style problem caught by buildifier

freedomtan committed:
fix typos caught by sanity check

s/delegte/delegate/

karimnosseir commented on tensorflow/lite/examples/label_image/label_image.cc:
```diff
@@ -85,9 +85,21 @@ TfLiteDelegatePtrMap GetDelegates(Settings* s) {
+
+  if (s->hexagon_delegate) {
+    const std::string libhexagon_path("/data/local/tmp");
+    auto delegate = evaluation::CreateHexagonDelegate(libhexagon_path);
```
The function takes another param. please fix.

freedomtan replied:
oops, will do it. my original patch was made before the `evaluation::CreateHexagonDelegate()` change, 1fac666679eb4353513b226060c844aa5a1d78ec.

freedomtan replied:
done

freedomtan committed:
reflect evaluation::CreateHexagonDelegate() change

rebased and checked that we use new version of
evaluation::CreateHexagonDelegate()
